src/v/cloud_storage/remote.cc

```diff
@@ -1364,14 +1364,20 @@ ss::future<remote::list_result> remote::list_objects(
     // Gathers the items from a series of successful ListObjectsV2 calls
     cloud_storage_clients::client::list_bucket_result list_bucket_result;
 
+    // If max_keys is 0, it is assumed the caller will handle truncation.
+    // max_keys will be passed as std::nullopt to client->list_objects()
+    // to allow for provider default max_keys.
+    const bool caller_handle_truncation
+      = (max_keys.has_value() && (max_keys.value() == 0));
+
```

Comment:
[Per the discussion with Nicolae](https://github.com/redpanda-data/redpanda/pull/18193#discussion_r1589232122), if `max_keys` is given a value of 0, we want to treat it as a special value that indicates the user wants to use the cloud provider default `max_keys` for the request, but not to fetch all objects in the event that the result is truncated (return after one round of requests).

Adding a `allow_truncation` flag and `vassert()`ing that `max_keys > 0` instead is also a very viable option. I think it's more explicit, and doesn't require the cloud provider default `max_keys` be used for the request.

What do you think?

Reply:
IMO it's more palatable and less error prone to be explicit, cc @nvartolomei since the special value is your suggestion

Reply:
@andrwng my thought process behind this proposal is this:

- If the caller passes `max_keys` then it implicitly states that it is ready to handle truncation or that it doesn't care about truncation. What would `allow_truncation=false` (default value?) along with `max_keys` do? 
  - a) Throw? Why would it throw if the caller asked for truncation via `max_keys` limit.
  - b) Not throw? Well, why wouldn't it? The caller didn't specify that truncation is allowed.

  Both outcomes are somewhat surprising/None of the outcomes is better than the other.

Not having an `allow_truncation` flag avoids this ambiguity. If `max_keys` is set, the caller implies that it knows how to handle truncation. The only question then is how the caller can communicate that is can handle truncation but that it also should get the maximum number of results that the backend can return? `std::numeric_limits<size_t>::max()`?

I suggested to use `0`, because my understanding was that `max_keys > 0` is actually interpreted as "minimum number of keys to return".

Maybe we should revisit that too.

- If max_keys is not specified, we collect all results under the assumption that the caller didn't implement truncation handling. It implicitly communicates "truncation is not allowed".
- If max_keys is specified, we propagate it to the cloud provider and always† return after the first response. It implicitly communicates that "truncation is allowed".

† I can't convince myself that interpreting max_keys as "iterate until end or at least this many results (min keys)" 
is beneficial under any circumstance. If the caller did specify max_keys, it knows how to handle truncation so let it handle it. Why do more allocations at this point? There is probably a use-case that the caller will want to do something like "list objects -> do something with them -> checkpoint" and bigger batches might be desirable before "checkpoint"-ing but that is probably an edge case we can ignore for now. The caller can implement that optimization. If this optimization becomes important we can later introduce ... min_keys.

If we make this change, then special treatment for `max_keys=0` won't be needed and we simplify the API even further.

Thoughts?

Reply:
- What happens if allow_truncation=false and max_keys is set?
---
- Separately, what are the cons of going via the following route? It reduces the API surface drastically and sounds like no surprises to me.

> - If max_keys is not specified, we collect all results under the assumption that the caller didn't implement truncation handling. It implicitly communicates "truncation is not allowed".
> - If max_keys is specified, we propagate it to the cloud provider and always† return after the first response. It implicitly communicates that "truncation is allowed".
---
> If the parameters of my request ultimately generates fewer keys than that, fine! 

That's exactly why I came up with the suggestion above. The caller stated that they know how to handle truncation so let them handle it. Don't do any magic inside as there is a very limited upside for doing that and then you have to come up with allow_truncation flag, tristate, etc...

Reply:
I think that your proposition is sound. We should communicate in the code the details around the possible case in which `max_keys >  cloud_provider_default_max_keys`:

- `max_keys`, if specified in a call to `list_objects()`, will be capped to the cloud provider default (which can vary).
- Anywhere `max_keys` is specified in a call to `list_objects()`, the user is responsible for checking the result for truncation, and to perform their own book-keeping at the call-site (probably in some sort of `while` loop) to ensure they obtain the desired number of object keys (i.e if they want `1001` keys, first call to `list_objects()` using `max_keys = 1001` returns `1000` objects due to cloud provider default, next call to `list_objects()` must provide `next_continuation_token` and `max_keys = 1` to get intended list of `1001` objects).

Reply:
Thanks for the rationale Nicolae.

I like the new proposal, I agree its semantics feel more intuitive.

Reply:
Changed behaviour around truncation and `max_keys` to Nicolae's proposal.